File: ios/Sasquatsh/Services/ChatService.swift
import Foundation

protocol ChatServiceProtocol {
    func getMessages(contextType: String, contextId: String, limit: Int, before: String?) async throws -> [ChatMessage]
    func sendMessage(contextType: String, contextId: String, content: String) async throws -> ChatMessage
    func deleteMessage(contextType: String, contextId: String, messageId: String) async throws
    func reportMessage(contextType: String, contextId: String, messageId: String, reason: String, details: String?) async throws
}

final class ChatService: ChatServiceProtocol {
    private let api: APIClient

    init(api: APIClient) {
        self.api = api
    }

    func getMessages(contextType: String, contextId: String, limit: Int = 50, before: String? = nil) async throws -> [ChatMessage] {
        var queryItems = [
            URLQueryItem(name: "contextType", value: contextType),
            URLQueryItem(name: "contextId", value: contextId),
            URLQueryItem(name: "limit", value: String(limit))
        ]
        if let before {
            queryItems.append(URLQueryItem(name: "before", value: before))
        }
        let response: ChatMessagesResponse = try await api.get("chat", queryItems: queryItems, authenticated: true)
        return response.messages
    }

    func sendMessage(contextType: String, contextId: String, content: String) async throws -> ChatMessage {
        let queryItems = [
            URLQueryItem(name: "contextType", value: contextType),
            URLQueryItem(name: "contextId", value: contextId)
        ]
        let input = SendMessageInput(content: content)
        let response: SendMessageResponse = try await api.post("chat", body: input, queryItems: queryItems, authenticated: true)
        return response.message
    }

    func deleteMessage(contextType: String, contextId: String, messageId: String) async throws {
        let queryItems = [
            URLQueryItem(name: "contextType", value: contextType),
            URLQueryItem(name: "contextId", value: contextId),
            URLQueryItem(name: "messageId", value: messageId)
        ]
        try await api.deleteVoid("chat", queryItems: queryItems, authenticated: true)
    }

    func reportMessage(contextType: String, contextId: String, messageId: String, reason: String, details: String?) async throws {
        let queryItems = [
            URLQueryItem(name: "contextType", value: contextType),
            URLQueryItem(name: "contextId", value: contextId),
            URLQueryItem(name: "messageId", value: messageId),
            URLQueryItem(name: "action", value: "report")
        ]
        let input = ReportMessageInput(reason: reason, details: details)
        let _: ReportMessageResponse = try await api.post("chat", body: input, queryItems: queryItems, authenticated: true)
    }
}

File: android/app/src/main/java/com/sasquatsh/app/services/ChatService.kt
package com.sasquatsh.app.services

import com.sasquatsh.app.models.*
import com.sasquatsh.app.services.api.ChatApi
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ChatService @Inject constructor(
    private val chatApi: ChatApi,
    private val moshi: Moshi
) {

    suspend fun getMessages(
        contextType: String,
        contextId: String,
        limit: Int,
        before: String?
    ): List<ChatMessage> {
        val response = chatApi.getMessages(contextType, contextId, limit, before)
        if (!response.isSuccessful) throw Exception("Failed to load messages")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        val wrapper = moshi.adapter(ChatMessagesResponse::class.java).fromJson(json)
        return wrapper?.messages ?: emptyList()
    }

    suspend fun sendMessage(
        contextType: String,
        contextId: String,
        content: String
    ): ChatMessage {
        val input = SendMessageInput(content = content)
        val response = chatApi.sendMessage(contextType, contextId, input)
        if (!response.isSuccessful) throw Exception("Failed to send message")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        val wrapper = moshi.adapter(SendMessageResponse::class.java).fromJson(json)
        return wrapper?.message ?: throw Exception("Failed to parse message")
    }

    suspend fun deleteMessage(
        contextType: String,
        contextId: String,
        messageId: String
    ) {
        val response = chatApi.deleteMessage(contextType, contextId, messageId)
        if (!response.isSuccessful) throw Exception("Failed to delete message")
    }

    suspend fun reportMessage(
        contextType: String,
        contextId: String,
        messageId: String,
        reason: String,
        details: String?
    ) {
        val input = ReportMessageInput(reason = reason, details = details)
        val response = chatApi.reportMessage(contextType, contextId, messageId, input = input)
        if (!response.isSuccessful) throw Exception("Failed to report message")
    }
}
